### PR TITLE
19291 unit test liquid templates

### PR DESCRIPTION
**Pull request by holdenhinkle**
## Description
Resolves [ticket #21258
](https://app.zenhub.com/workspaces/vsp-5cedc9cce6e3335dc5a49fc4/issues/department-of-veterans-affairs/va.gov-team/21258)
This PR creates an MVP of a Liquid Template Unit Testing Framework. Includes updated documentation.

## Testing done
Several example tests are included. Run:
- `yarn test:unit src/site/layouts/tests/vamc/health_care_region_page.unit.spec.js`
- `yarn test:unit src/site/layouts/tests/landing_page/landing_page.unit.spec.js`

## Acceptance criteria
 - [x] Define a minimal framework that supports the core functions of the liquid template implementation on the platform
 - [x] Verify that the framework is capable of testing some representative subset of liquid templates
 - [x] Work with stakeholders to verify that the MVP fulfills the core of their requirements

## Definition of done
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: src/site/tests/support/index.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+import registerFilters from '../../filters/liquid.js';
+import createRedirects from '../../stages/build/plugins/rewrite-va-domains.js';
+import rewriteAWSUrls from '../../stages/build/plugins/rewrite-cms-aws-urls.js';
+import modifyDom from '../../stages/build/plugins/modify-dom';
+
+registerFilters();
+
+const updateHTML = files => {
+  const options = {
+    buildtype: process.env.BUILDTYPE, // what should i set this to?
```
Comment (mpelzsherman) [inline src/site/tests/support/index.js:14]: We can probably just hardcode this to `ENVIRONMENTS.VAGOVDEV`

`import ENVIRONMENTS from 'site/constants/environments';`

File: src/site/tests/support/index.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    // what should the following values be?
```
Comment (mpelzsherman) [inline src/site/tests/support/index.js:15]: What you have here is fine AFAIC.

**Comment by holdenhinkle** [comment]
To test this using the two test examples, run:
- `yarn test:unit src/site/layouts/tests/vamc/health_care_region_page.unit.spec.js`
- `yarn test:unit src/site/layouts/tests/landing_page/landing_page.unit.spec.js`

File: src/site/layouts/tests/landing_page/landing_page.unit.spec.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+      const container = await renderHTML(layoutPath, layout, data);
+      expect(container.querySelector('h1').innerHTML).to.equal(data.title);
+      expect(container.querySelector('.va-introtext p').innerHTML).to.equal(
+        data.fieldIntroText,
+      );
+      expect(
+        container.querySelector('i.icon-large.white.hub-icon-foo'),
+      ).not.to.equal(null);
+    });
+  });
```
Comment (ncksllvn) [inline src/site/layouts/tests/landing_page/landing_page.unit.spec.js:42]: Imagine a utility like `executeAxeCheck` -> https://github.com/department-of-veterans-affairs/vets-website/pull/10623/files#diff-fca242d8f94a9b0add72c42c8676727b81504cd1f52eb8a4e898c6992807bcf0R32

```suggestion
  });
  it('passes an axe check-, () => {
    const container = await renderHTML(layoutPath, layout, data);
    executeAxeCheck('https://www.va.gov/something', container);
  });
```

**Review by mpelzsherman — APPROVED** [review]

File: src/site/README.md [hunk 4]
```diff
@@ -307,3 +307,36 @@ The following query looks up the `entityId` for one page of the `entityBundle` (
 ```
 
 </details>
+
+## Testing
```
Comment (timwright12) [inline src/site/README.md:311]: @holdenhinkle Any reason Cypress wasn't used for these tests?

File: src/site/stages/build/plugins/modify-dom/index.js [hunk 5]
```diff
@@ -8,16 +8,29 @@ const addSubheadingsIds = require('./add-id-to-subheadings');
...
-const domModifiers = [
-  addNonceToScripts,
-  processEntryNames,
-  updateExternalLinks,
-  addSubheadingsIds,
-  checkBrokenLinks,
-  injectAxeCore,
-];
+const getDomModifiers = BUILD_OPTIONS => {
+  if (BUILD_OPTIONS.buildtype === 'vagovdev') {
```
Comment (timwright12) [inline src/site/stages/build/plugins/modify-dom/index.js:12]: Any reason addNonceToScripts and checkBrokenLinks were remove here for dev builds? Is this related to the tests or hitching a ride?

File: src/site/README.md (hunk 4 above)
Comment (holdenhinkle) [reply src/site/README.md]: Hey @timwright12 , I just added a 'why not cypress' note to the documentation.

File: src/site/stages/build/plugins/modify-dom/index.js (hunk 5 above)
Comment (holdenhinkle) [reply src/site/stages/build/plugins/modify-dom/index.js:12]: The `addNonceToScripts` and `checkBrokenLinks` `domModifiers` are being excluded because they're not needed for the unit tests.

File: src/site/README.md [hunk 6]
```diff
@@ -307,3 +307,39 @@ The following query looks up the `entityId` for one page of the `entityBundle` (
 ```
 
 </details>
+
+## Liquid Template Unit Testing Framework
+
+The Liquid Template Unit Testing Framework was created to replace Cypress, which is a slow and heavy framework and is overkill for testing logic in `liquid` templates.
+
```
Comment (mpelzsherman) [inline src/site/README.md:314]: Cypress tests do not give us control over the test data, which is crtitical for these tests.

Comment (timwright12) [reply src/site/README.md:314]: Can we elaborate on "The Liquid Template Unit Testing Framework was created to replace Cypress" is this a longterm update plan on the site? Last I heard some teams are still trying to migrate from Nightwatch to Cypress

Also, should we be adding this type of documentation to:

- https://department-of-veterans-affairs.github.io/va.gov-team/
- https://department-of-veterans-affairs.github.io/veteran-facing-services-tools/getting-started/common-tasks/test

Rather than in the repo like this?

File: src/site/stages/build/plugins/modify-dom/index.js (hunk 5 above)
Comment (holdenhinkle) [reply src/site/stages/build/plugins/modify-dom/index.js:12]: I was discussing the best way to flag those two modifiers as not-needed with Nick Sullivan and he suggested; “maybe check `BUILD_OPTIONS.BUILDTYPE` and skip if it’s `vagovdev`?”

Alternatively, I can add a property to the `options` that are passed in called something like ‘unit_testing_framework’ and set it to `true`?

Comment (holdenhinkle) [reply src/site/stages/build/plugins/modify-dom/index.js]: I updated the `BUILD_OPTIONS` to:
```
  const options = {
    buildtype: BUILDTYPE,
    entry: true,
    liquidUnitTestingFramework: true,
    // the following options are expected but the values don't matter
    host: 'host',
    port: 3001,
    protocol: 'http',
  };
```

and set the `domModifiers` array using `liquidUnitTestingFramework` instead of `buildtype`:
```
const getDomModifiers = BUILD_OPTIONS => {
  if (BUILD_OPTIONS.liquidUnitTestingFramework) {
    return [
      processEntryNames,
      updateExternalLinks,
      addSubheadingsIds,
      injectAxeCore,
    ];
  }

  return [
    addNonceToScripts,
    processEntryNames,
    updateExternalLinks,
    addSubheadingsIds,
    checkBrokenLinks,
    injectAxeCore,
  ];
};
```

File: src/site/README.md (hunk 6 above)
Comment (holdenhinkle) [reply src/site/README.md:314]: I'll add documentation to `https://department-of-veterans-affairs.github.io/va.gov-team/` in a fast-follow pr.

Comment (timwright12) [reply src/site/README.md:314]: This still reads like Cypress is being completely killed off, is that the case? If so I think that needs to be communicated to teams currently adding Cypress tests

Comment (holdenhinkle) [reply src/site/README.md:314]: I updated that paragraph to:

```
The Liquid Template Unit Testing Framework was created to replace Cypress for unit testing the logic in `liquid` templates because Cypress is slow and heavy and is overkill for this purpose. Please note: End-to-end (e2e) tests on VA.gov use Cypress; Cypress has not been replaced by this tool.
```

**Review by timwright12 — APPROVED** [review]
@holdenhinkle Largely I'm good with this if it's solving a problem we have and not adding more flakiness. Very strongly feel that we need to not have this documentation here, but rather in the official docs, will look for that PR. Can you ping me when you have it ready? I'd like to read through it for my own edification.

It would be great to layout some "dos" and "dont's" for this to avoid any potential flakiness.

**Comment by holdenhinkle** [comment]
> @holdenhinkle Largely I'm good with this if it's solving a problem we have and not adding more flakiness. Very strongly feel that we need to not have this documentation here, but rather in the official docs, will look for that PR. Can you ping me when you have it ready? I'd like to read through it for my own edification.
> 
> It would be great to layout some "dos" and "dont's" for this to avoid any potential flakiness.

Thanks Tim. Yes, I'll for sure ping you with a link to the addition to the official docs.
